crypto/objects/o_names.c

```diff
@@ -157,11 +174,13 @@ const char *OBJ_NAME_get(const char *name, int type)
 {
     OBJ_NAME on, *ret;
     int num = 0, alias;
+    const char *value = NULL;
 
     if (name == NULL)
         return (NULL);
-    if ((names_lh == NULL) && !OBJ_NAME_init())
+    if (!OBJ_NAME_init())
         return (NULL);
+    CRYPTO_THREAD_read_lock(lock);
```

Comment:
FWIW, there's other code which takes a read lock around `lh_*_retrieve`, like `ssl_get_prev_session` and `int_err_get_item`. But there's also performance reasons why you'd probably not want a write lock around those. I would have suggested taking those stats out, but sadly they're public API by way of `OPENSSL_LH_stats_bio`...